planner/core/exhaust_physical_plans.go

```diff
@@ -387,6 +387,93 @@ func (p *LogicalJoin) constructIndexJoin(
 	return []PhysicalPlan{join}
 }
 
+func findIndexScanAndTableScan(p PhysicalPlan) (is *PhysicalIndexScan, ts *PhysicalTableScan) {
+	var children []PhysicalPlan
+	switch x := p.(type) {
+	case *PhysicalTableReader:
+		children = x.TablePlans
+	case *PhysicalIndexReader:
+		children = x.IndexPlans
+	case *PhysicalIndexLookUpReader:
+		children = append(x.IndexPlans, x.TablePlans...)
+	default:
+		children = x.Children()
+	}
+	for _, child := range children {
+		tmpIS, tmpTS := findIndexScanAndTableScan(child)
+		if tmpIS != nil {
+			is = tmpIS
+		}
+		if tmpTS != nil {
+			ts = tmpTS
+		}
+	}
+	if tmpIS, ok := p.(*PhysicalIndexScan); ok {
+		return tmpIS, ts
+	}
+	if tmpTS, ok := p.(*PhysicalTableScan); ok {
+		return is, tmpTS
+	}
+	return is, ts
+}
+
+func (p *LogicalJoin) constructIndexMergeJoin(
+	prop *property.PhysicalProperty,
+	outerIdx int,
+	innerPlan PhysicalPlan,
+	ranges []*ranger.Range,
+	keyOff2IdxOff []int,
+	lens []int,
+	compareFilters *ColWithCmpFuncManager,
+) []PhysicalPlan {
+	indexJoins := p.constructIndexJoin(prop, outerIdx, innerPlan, ranges, keyOff2IdxOff, lens, compareFilters)
+	indexMergeJoins := make([]PhysicalPlan, 0, len(indexJoins))
+	for _, plan := range indexJoins {
+		is, ts := findIndexScanAndTableScan(innerPlan)
+		join := plan.(*PhysicalIndexJoin)
+		// needOuterSort means whether the outer join keys are the prefix of the prop items.
+		needOuterSort := len(prop.Items) > 0 && len(join.OuterJoinKeys) <= len(prop.Items)
+		compareFuncs := make([]expression.CompareFunc, 0, len(join.OuterJoinKeys))
+		outerCompareFuncs := make([]expression.CompareFunc, 0, len(join.OuterJoinKeys))
+		for i := range join.OuterJoinKeys {
+			if needOuterSort && !prop.Items[i].Col.Equal(nil, join.OuterJoinKeys[i]) {
+				needOuterSort = false
```

Comment:
Should we `break` once `needOuterSort` is set to `false`?

Reply:
No, we can't. If `break`, the `compareFuncs` can't be appended for all join keys.